README.md
# At_Base2<sup>15</sup>

binary-to-text encoding scheme that represents binary data in an unicode string format. Each unicode character represents 15 bits of binary data.

#### Example ####

| Encoding | Data | chararacters |
|:-:|:-:|:-:|
| Plain text | Base2e15 is awesome! | 20 |
| **Base2<sup>15</sup>** | **嗺둽嬖蟝巍媖疌켉溁닽壪** | **11** |
| Base64 | QmFzZTJlMTUgaXMgYXdlc29tZSE= | 27+1 |
 
## Mapping table

Every character represents 15 bits of data, except the last character, which is 7 or 15 bits.

| Binary | Unicode | Unicode Range Name |
|:-:|:-:|:-:|
| **15 bits mapping** | | |
| 0x0000 ~ 0x1935 | U+3480 ~ U+4DB5 | CJK Unified Ideographs Extension A |
| 0x1936 ~ 0x545B | U+4E00 ~ U+8925 | CJK Unified Ideographs |
| 0x545C ~ 0x7FFF | U+AC00 ~ U+D7A3 | Hangul Syllables |
| **7 bits mapping** | | |
| 0x00   ~ 0x7F | U+3400 ~ U+347F | CJK Unified Ideographs Extension A |

## Usage

A simple usage example in dart:
```dart
import 'dart:convert';
import 'package:base2e15/base2e15.dart';

main() {
  String msg = 'Base2e15 is awesome!';
  String encoded = Base2e15.encode(UTF8.encode(msg));
  String decoded = UTF8.decode(Base2e15.decode(encoded));
}
```

## Compare

| Compare | Base2<sup>15</sup> |  Base64 |
|:-:|:-:|:-:|
| bits per character | **15** | 6 |
| bits per char width | **7.5 (15/2)** | 6 (6/1) |
| bits per UTF8 byte | 5 (15/3) | **6 (6/1)** |
| bits per UTF16 byte | **7.5 (15/2)** | 3 (6/2) |

## Why not Base2<sup>16</sup>?

The unicode range `CJK Unified Ideographs Extension B` contains 42711 characters (U+20000 ~ U+2A6D6), together with the characters used by Base2<sup>15</sup>, there are more than 65536 usable characters to encode 16 bits in each character.

However, font support for `CJK Unified Ideographs Extension B` is missing in most mobile devices and using this code range will also reduce the bits capacity in UTF8 and UTF16 encoding, since those characters require one more byte in UTF8 and 2 more bytes in UTF16.

## Why is this forked from [rinick/base2e15](https://github.com/rinick/base2e15)?

We use base2e15 as an underlying dependency for a number of things, and need
a null safety version so that the higher order dependencies can offer null
safety.

## What about other language support?

Support for C, C# & Java has been removed from this fork as we're only
interested in Dart null safety.

dart/CHANGELOG.md
# Changelog

## 1.0.0

- Initial version



dart/test/all_test.dart
// Copyright (c) 2015, Rick Zhou. All rights reserved. Use of this source code
// is governed by a BSD-style license that can be found in the LICENSE file.

library at_base2e15.test;

import 'package:at_base2e15/at_base2e15.dart';
import 'dart:convert';
//import 'dart:math';

bool testEqual(Object a, Object b, String testName) {
  if (a == b) {
    print('$testName Passed!');
    return true;
  } else {
    print('$testName Failed, "$a" != "$b"');
    return false;
  }
}

main() {
  String msg = 'Base2e15 is awesome!';
  String encoded = Base2e15.encode(utf8.encode(msg));
  testEqual(encoded, '嗺둽嬖蟝巍媖疌켉溁닽壪', 'Encoding Test');
  String decoded = utf8.decode(Base2e15.decode(encoded));
  testEqual(decoded, msg, 'Decoding Test');
  String encoded2 = '~嗺둽嬖蟝巍媖疌123켉溁닽壪';
  String decoded2 = utf8.decode(Base2e15.decode(encoded2));
  testEqual(decoded2, msg, 'Malformed Decoding Test');

//  Random rng = new Random();
//  List bytes = new List(100);
//  for (int i = 0; i < 10000; ++i) {
//    int k = 50 + rng.nextInt(50);
//    for (int j = 0; j < k; ++j) {
//      bytes[j] = rng.nextInt(256);
//    }
//    String encoded = Base2e15.encode(bytes.sublist(0,k));
//    List newbytes = Base2e15.decode(encoded);
//    for (int j = 0; j < k; ++j) {
//      if (bytes[j] != newbytes[j]) {
//        print('Failed:$bytes');
//        print('      :$newbytes');
//        break;
//      }
//    }
//  }
}


dart/lib/at_base2e15.dart
// Copyright (c) 2015, Rick Zhou. All rights reserved. Use of this source code
// is governed by a BSD-style license that can be found in the LICENSE file.

/// The base2e15 library.
/// Map 15 bits to unicode
/// 0x0000 ~ 0x1935 -> U+3480 ~ U+4DB5   CJK Unified Ideographs Extension A
/// 0x1936 ~ 0x545B -> U+4E00 ~ U+8925   CJK Unified Ideographs
/// 0x545C ~ 0x7FFF -> U+AC00 ~ U+D7A3   Hangul Syllables
/// 7 bits special case, only used by last character
///  0x00  ~  0x7F  -> U+3400 ~ U+347F   CJK Unified Ideographs Extension A
library at_base2e15;

import 'dart:typed_data';

class Base2e15 {
  static String encode(List<int> bytes,
      [int lineSize = 0, String? linePadding]) {
    List<int?> charCodes = encodeToCharCode(bytes);
    if (lineSize <= 0) {
      return new String.fromCharCodes(charCodes as Iterable<int>);
    }
    List rslt = [];
    int len = charCodes.length;
    for (int i = 0; i < len; i += lineSize) {
      int j = i + lineSize;
      if (j < len) {
        j = len;
      }
      if (linePadding != null) {
        rslt.add(
            '$linePadding${new String.fromCharCodes(charCodes.sublist(i, j) as Iterable<int>)}');
      } else {
        rslt.add(
            new String.fromCharCodes(charCodes.sublist(i, j) as Iterable<int>));
      }
    }
    return rslt.join('\n');
  }

  static List<int?> encodeToCharCode(List<int> bytes) {
    int bn = 15; // bit needed
    int bv = 0; // bit value
    int outLen = (bytes.length * 8 + 14) ~/ 15;
    List<int> out = new List<int>.filled(outLen, -1, growable: false);
    int pos = 0;
    for (int byte in bytes) {
      if (bn > 8) {
        bv = (bv << 8) | byte;
        bn -= 8;
      } else {
        bv = ((bv << bn) | (byte >> (8 - bn))) & 0x7FFF;
        if (bv < 0x1936) {
          out[pos++] = bv + 0x3480;
        } else if (bv < 0x545C) {
          out[pos++] = bv + 0x34CA;
        } else {
          out[pos++] = bv + 0x57A4;
        }
        bv = byte;
        bn += 7;
      }
    }
    if (bn != 15) {
      if (bn > 7) {
        // need 8 bits or more, so has 7 bits or less
        out[pos++] = ((bv << (bn - 8)) & 0x7F) + 0x3400;
      } else {
        bv = (bv << bn) & 0x7FFF;
        if (bv < 0x1936) {
          out[pos++] = bv + 0x3480;
        } else if (bv < 0x545C) {
          out[pos++] = bv + 0x34CA;
        } else {
          out[pos++] = bv + 0x57A4;
        }
      }
    }
    return out;
  }

  static Uint8List decode(String input) {
    int bn = 8; // bit needed
    int bv = 0; // bit value
    int maxLen = (input.length * 15 + 7) ~/ 8;
    Uint8List out = new Uint8List(maxLen);
    int pos = 0;
    int cv;
    for (int code in input.codeUnits) {
      if (code > 0x33FF && code < 0xD7A4) {
        if (code > 0xABFF) {
          cv = code - 0x57A4;
        } else if (code > 0x8925) {
          continue; // invalid range
        } else if (code > 0x4DFF) {
          cv = code - 0x34CA;
        } else if (code > 0x4DB5) {
          continue; // invalid range
        } else if (code > 0x347F) {
          cv = code - 0x3480;
        } else {
          cv = code - 0x3400;
          out[pos++] = (bv << bn) | (cv >> (7 - bn));
          break; // last 8 bit data received, break
        }
        out[pos++] = (bv << bn) | (cv >> (15 - bn));
        bv = cv;
        bn -= 7;
        if (bn < 1) {
          out[pos++] = bv >> -bn;
          bn += 8;
        }
      }
    }
    return out.sublist(0, pos);
  }
}
